### PR TITLE
WIP: LwM2M FW object - multiple instances

### PR DESCRIPTION
**Pull request by VeijoPesonen**
Dedicated instance for each firmware image found from the device
* Cellular modem firmware
* Application firmware

For further explanation of this standard incompliant change please see the issue OpenMobileAlliance/lwm2m-registry#615.

These changes can be tested by using the Leshan LwM2M server which allows multiple Firmware Update object instances.
image

Requires:  nrfconnect/sdk-zephyr#590

### CODE REVIEW

**Comment by rlubos** [comment]
Aren't these changes agains LwM2M spec? It makes it pretty clear, that LwM2M Firmware Update is a single-instance object.

**Comment by VeijoPesonen** [comment]
> Aren't these changes against LwM2M spec? It makes it pretty clear, that LwM2M Firmware Update is a single-instance object.

@rlubos Yeah, you're absolutely right. We are considering if it would be worth of trying to get the standard updated. That might be a long road. Next thing you might be asking is that why not use the [LwM2M Software Management object ](https://www.md2html-tool.com/docs/OpenMobileAlliance/SwMgmt/dev/2b37235/SUP/9.html)- Object ID 9 - but that's overly complicated.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 1]
```diff
@@ -307,15 +361,18 @@ int lwm2m_init_image(void)
 		}
 	}
 
+	char path[10+1]; /* '/5/[0-65535]/5' -> max len 10 + terminating NULL */
```
[the remark this answers is not in the document]
Comment (VeijoPesonen) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c]: Good catch. Fixed.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 2]
```diff
@@ -307,15 +361,18 @@ int lwm2m_init_image(void)
...
+
+	ret = snprintf(path, sizeof(path), "/5/%" PRIu16 "/0", obj_inst_id);
```
[the remark this answers is not in the document]
Comment (VeijoPesonen) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c]: I don't know what I have been thinking here. Fixed.

File: subsys/net/lib/lwm2m_client_utils/fota/settings.c [hunk 3]
```diff
@@ -7,54 +7,76 @@
...
-int fota_update_counter_read(struct update_counter *update_counter)
+static struct update_counter uc[MAX_INSTANCE_COUNT];
+
+static const char fctr_path[] = "fota/counter";
+static const char ctr_str[] = "counter";
+
+int fota_update_counter_read(uint16_t obj_inst_id, struct update_counter *update_counter)
 {
-	memcpy(update_counter, &uc, sizeof(uc));
+	memcpy(update_counter, &uc, sizeof(uc[obj_inst_id]));
```
[the remark this answers is not in the document]
Comment (VeijoPesonen) [reply subsys/net/lib/lwm2m_client_utils/fota/settings.c]: You're absolutely right. Fixed.

File: include/net/lwm2m_client_utils_fota.h [hunk 4]
```diff
@@ -33,13 +33,16 @@ enum counter_type {
 
 /**
  * @brief Read the update counter
+ *
+ * @param[in] obj_inst_id Firmware Object instance ID, only standardized value is zero
+ * @param[out] update_counter values for active and
```
[the remark this answers is not in the document]
Comment (VeijoPesonen) [reply include/net/lwm2m_client_utils_fota.h]: Fixed.

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 5]
```diff
@@ -47,6 +47,26 @@ config LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPORT
...
 
+if LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPORT
+
+config LWM2M_CLIENT_UTILS_FIRMWARE_INSTANCE_MODEM
+	int "Modem Firmware Object instance ID, highest in the range is not usable"
+	# Neither arithmetic operations are not possible when defining a range,
+	# nor excluding the values which equal to second symbol - or the first to that matter;
+	# it's impossibile to limit the range of the value to LWM2M_FIRMWARE_INSTANCE_COUNT-1.
+	range 0 LWM2M_FIRMWARE_INSTANCE_COUNT
+	default 0
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:58]: Move these explanations regarding the range to "help" section of Kconfig.

```
config LWM2M_CLIENT_UTILS_FIRMWARE_INSTANCE_MODEM
	int "Modem Firmware Object instance ID"
	range 0 LWM2M_FIRMWARE_INSTANCE_COUNT
	default 0
	help
          Instance ID must be in range from 0 to LWM2M_FIRMWARE_INSTANCE_COUNT-1.
          Highest in the range is not usable.
```

Comment (VeijoPesonen) [reply subsys/net/lib/lwm2m_client_utils/Kconfig:55]: Actually this is what you will see - the help section seems to get omitted. I placed the earlier section marked as comment to _help_ but it's not show anywhere.

image

Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m_client_utils/Kconfig:55]: Am I missing something, as I got a build error from this config:
```
warning: the int symbol LWM2M_CLIENT_UTILS_FIRMWARE_INSTANCE_MODEM (defined at /home/seta/src/ncs/nrf/subsys/net/lib/lwm2m_client_utils/Kconfig:52) has a non-int range [0 (undefined), LWM2M_FIRMWARE_INSTANCE_COUNT (undefined)]

warning: the int symbol LWM2M_CLIENT_UTILS_FIRMWARE_INSTANCE_APP (defined at /home/seta/src/ncs/nrf/subsys/net/lib/lwm2m_client_utils/Kconfig:62) has a non-int range [0 (undefined), LWM2M_FIRMWARE_INSTANCE_COUNT (undefined)]

/home/seta/src/ncs/nrf/samples/nrf9160/lwm2m_client/prj.conf:36: warning: attempt to assign the value '2' to the undefined symbol LWM2M_FIRMWARE_INSTANCE_COUNT
```

Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m_client_utils/Kconfig:55]: The help text is visible if you press `?`.
But I think it is OK to include the warning in both.

2021-09-10 lwm2m_config

**Comment by VeijoPesonen** [comment]
Not needed based on discussion under OpenMobileAlliance/lwm2m-registry#638.

**Pull request closed without merging**